// format-plan-pdf/format-plan-pdf.js
'use strict';

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	//LOCAL VARIABLES ====================================

	const pdfLibrary = new args.pdfLibrary(args)

	const buildHeaderSection = new args.buildHeaderSection({
		pdfLibrary: pdfLibrary
	})
	const buildStudentSection = new args.buildStudentSection({
		pdfLibrary: pdfLibrary
	})
	const buildInfoSection = new args.buildInfoSection({
		pdfLibrary: pdfLibrary
	})
	const buildPlanSection = new args.buildPlanSection({
		pdfLibrary: pdfLibrary
	})

	const pdfMake=args.pdfMake;
	const assembleTextCell=pdfLibrary.exportPackage.assembleTextCell;

	//CONSTRUCTION ====================================

	pdfLibrary.registerStyleDefinition('container', {
			margin: [0, 0, 0, 0]
		});

	pdfLibrary.registerStyleDefinition('pageTitle', {
			bold: true,
			fontSize: 22
		});

	const assembleReportBody = function() {

		const section1 = buildHeaderSection.docSpec();
		const section2 = buildStudentSection.docSpec();
		const section3 = buildInfoSection.docSpec();
		const section4 = buildPlanSection.docSpec();

		const body = []

		body.push([assembleTextCell('table', section1, '', '', '', 'center')])
		body.push([assembleTextCell('table', section2, '', '', '', 'center')])
		body.push([assembleTextCell('table', section3, '', '', '', 'center')])
		body.push([assembleTextCell('table', section4, '', '', '', 'center')])

		return body;

	};

	const assembleContent = function() {
		return  [
			{
				style: 'container',
				layout: 'noBorders',
				table: {
					widths: ['*'],
					body: assembleReportBody()
				}
			},
		];
	}

	const finalDocSpec = {
		background: [{
			text: 'CARE PLANNER',
			color: 'gray'
		},
		],
		pageOrientation: 'landscape',
		pageMargins: [30, 30, 30, 30],
		defaultStyle: {
			fontSize: 10
		},
		content: assembleContent(),
		images: {},
		styles: pdfLibrary.namedStyleDefinitions()

	}

	//METHODS AND PROPERTIES ====================================

	this.getDataUrl = function(callback) {
		const pdfDocGenerator = pdfMake.createPdf(finalDocSpec);
		pdfDocGenerator.getDataUrl(callback);
	}

	//API ENDPOINTS ====================================

	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
