// format-plan-pdf/format-plan-pdf.js
'use strict';

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	const qtools = args.qtools;
	const pdfMake = args.pdfMake;
	const student = args.student;
	const plan = args.plan;

	const self = this;


	//LOCAL VARIABLES ====================================

	//LOCAL FUNCTIONS ====================================

	const getSimplePlanString = function() {

		let outArray = [];

		var addToString = function(item, type) {
			if (!item) {
				return;
			}
			let format = [110, 2, 10, 20];
			switch (type) {
				case 'condition':
					format = [115, 2, 10, 20];
					outArray.push({
						text: item,
						margin: format,
						fontSize: 15,
						bold: true
					});
					break;
				case 'diagnosis':
					format = [215, 2, 10, 20];
					outArray.push({
						text: item,
						margin: format
					});
					break;
				case 'shortName':
					format = [215, 2, 10, 20];
					outArray.push({
						text: item,
						margin: format,
						bold: true,
						italics: true
					});
					break;

			}


		}

		for (var i = 0, len = plan.conditions.length; i < len; i++) {
			var condition = plan.conditions[i];
			addToString(condition.title, 'condition');


			for (var j = 0, len2 = condition.diagnoses.length; j < len2; j++) {
				var diagnosis = condition.diagnoses[j];

				addToString(diagnosis.shortName, 'shortName');
				addToString(diagnosis.assessmentData, 'diagnosis');
				addToString(diagnosis.nursingDiagnosis, 'diagnosis');
				addToString(diagnosis.interventions, 'diagnosis');
				addToString(diagnosis.outcomes, 'diagnosis');


			}
			addToString('');
			addToString('');
		}



		console.dir({
			"outArray": outArray
		});



		return outArray;
	}


	//METHODS AND PROPERTIES ====================================


	const documentObject = {
		background: 'CARE PLANNER',
		pageOrientation: 'landscape',
		pageMargins: [40, 60, 40, 60],
		defaultstyle: {},
		content: [],
		images:{},
		styles:{}

	}
	this.getDataUrl = function(callback) {

		//var docDefinition = { content: `Student: ${student.first} ${student.last}\n\n${getSimplePlanString()}` };


		documentObject.content = getSimplePlanString();

		const pdfDocGenerator = pdfMake.createPdf(documentObject);
		pdfDocGenerator.getDataUrl(callback);


	}

	//API ENDPOINTS ====================================

	//INITIALIZATION ====================================

	console.log(__dirname);

	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
